### DISCUSSION TITLE
`await browser.getPuppeteer()` is not working in Firefox tests running

### DISCUSSION THREAD

**Discussion by VladSelikh** [webdriverio/webdriverio#12769] [Q&A] [answered]
Hey guys!
I receive the error while running the command `await browser.getPuppeteer()` in my tests. I need a Puppeteer browser to access the clipboard value. The Error text is:
`TypeError [ERR_INVALID_URL]: Invalid URL`
I've set `'moz:debuggerAddress'` as `true` in my caps.
Could you please advise whether I can run this command and use Puppeteer browser for Firefox at all?
If yes, what am I doing wrong?
If no, is there any other way to access clipboard in Firefox?

**Comment by christian-bromann** [comment]
> I've set `'moz:debuggerAddress'` as `true` in my caps.

Why do you set the capability like that?

**Reply by VladSelikh** [reply]
Without this one set in `true` I still have the same Error, just thought that in this way I would foresee the suggestions to do that😁

**Reply by christian-bromann** [reply]
Are you running your automation script locally?

**Reply by VladSelikh** [reply]
Yes

**Reply by christian-bromann** [reply]
It seems like this is not supported by Firefox anymore, even though it was supported a while ago. We have a note to our command saying:

> Note that using Puppeteer requires support for Chrome DevTools protocol and e.g. can not be used when running automated tests in the cloud. Find out more in the [Automation Protocols](https://webdriver.io/docs/automationProtocols) section.

I think it makes sense to be clear about that this will only work when running Chrome tests locally.

**Reply by VladSelikh** [reply]
Got it, thanks😉

**Comment by VladSelikh** [comment, marked as answer]
I've found the solution of how I can access clipboard data in Firefox
I just needed to adjust Firefox caps in my config file:
```
{
    browserName: 'firefox',
    'moz:firefoxOptions': {
      prefs: {
        'dom.events.asyncClipboard.readText': true,
      }
    },
  };
```
...and then in tests just use:
```
const clipboardText = await browser.execute(() => navigator.clipboard.readText());
await browser.url(clipboardText);
```

**Reply by christian-bromann** [reply]
Thanks for sharing this! I raised https://github.com/webdriverio/example-recipes/issues/418 to make sure we eventually have better docs for this. Feel free to contribute 😉
